File: drills/elixir/solutions/nc146_validate_bst.exs
defmodule Solution do
  # Check against a range, not against the parent. A node can be larger than its
  # own parent and still break the order, because the constraint comes from an
  # ancestor further up -- and that is the whole difficulty. Going left tightens
  # the upper bound, going right the lower one.
  def is_valid_bst(tree), do: within(tree, nil, nil)

  defp within(nil, _low, _high), do: true

  defp within({value, left, right}, low, high) do
    above?(value, low) and below?(value, high) and
      within(left, low, value) and within(right, value, high)
  end

  defp above?(_value, nil), do: true
  defp above?(value, low), do: value > low

  defp below?(_value, nil), do: true
  defp below?(value, high), do: value < high
end

File: drills/go/solutions/nc146_validate_bst.go
package main

import "math"

func isValidBST(root *TreeNode) bool {
	// Every node must lie strictly inside the bounds its ancestors set:
	// going left tightens the upper bound, going right the lower.
	var valid func(node *TreeNode, low, high int) bool
	valid = func(node *TreeNode, low, high int) bool {
		if node == nil {
			return true
		}
		if node.Val <= low || node.Val >= high {
			return false
		}
		return valid(node.Left, low, node.Val) && valid(node.Right, node.Val, high)
	}
	return valid(root, math.MinInt, math.MaxInt)
}
